File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int left = 0;
        int right = height.size()-1;
        int leftMax = height[0];
        int rightMax = height[height.size()-1];
        
        int ans = 0;
        
        while (left < right){
            if (leftMax < rightMax){
                left++;
                leftMax = max(leftMax,height[left]);
                ans += (leftMax - height[left]);
            }
            else{
                right--;
                rightMax = max(rightMax,height[right]);
                ans += (rightMax - height[right]);
            }
        }
        
        return ans;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        
        left = 0
        right = len(height) - 1
        leftMax = height[left]
        rightMax = height[right]
        ans = 0

        while left < right:
            if leftMax < rightMax:
                left += 1
                leftMax = max(leftMax,height[left])
                ans += (leftMax - height[left])
            else:
                right -= 1
                rightMax = max(rightMax,height[right])
                ans += (rightMax - height[right])
                
        return ans
        
        